packages/salesforcedx-vscode-core/test/jest/commands/retrieveExecutor.test.ts

```diff
@@ -0,0 +1,98 @@
+import { WorkspaceContext } from '../../../src/context/workspaceContext';
+import { SourceTrackingService } from '../../../src/services';
+
+const dummyProjectPath = '/a/project/path';
+jest.mock('@salesforce/source-deploy-retrieve', () => {
+  return {
+    ...jest.requireActual('@salesforce/source-deploy-retrieve'),
+    getRootWorkspacePath: () => dummyProjectPath,
+    ComponentSet: jest.fn().mockImplementation(() => {
+      return {
+        retrieve: jest.fn().mockImplementation(() => {
+          return { pollStatus: jest.fn() };
+        })
+      };
+    })
+  };
+});
+
+jest.mock('@salesforce/salesforcedx-utils-vscode', () => {
+  return {
+    ...jest.requireActual('@salesforce/salesforcedx-utils-vscode'),
+    getRootWorkspacePath: () => dummyProjectPath,
+    ChannelService: jest.fn().mockImplementation(() => {
+      return {};
+    }),
+    TelemetryService: { getInstance: jest.fn() },
+    TelemetryBuilder: jest.fn()
+  };
+});
+
+jest.mock('../../../src/messages', () => {
+  return { loadMessageBundle: jest.fn(), nls: { localize: jest.fn() } };
+});
+
+jest.mock('../../../src/commands/util/postconditionCheckers');
+
+jest.mock('../../../src/sfdxProject/sfdxProjectConfig');
+
+describe('Retrieve Executor', () => {
+  const dummyProcessCwd = '/';
+  const workspaceContextGetInstanceSpy = jest.spyOn(
+    WorkspaceContext,
+    'getInstance'
+  );
+  const mockWorkspaceContext = { getConnection: jest.fn() } as any;
+  const createSourceTrackingSpy = jest.spyOn(
+    SourceTrackingService,
+    'createSourceTracking'
+  );
+  const dummyComponentSet = new ComponentSet();
+  const retrieveSpy = jest.spyOn(dummyComponentSet, 'retrieve');
+
+  class TestRetrieveExecutor extends RetrieveExecutor<{}> {
+    protected getComponents(
+      response: ContinueResponse<{}>
+    ): Promise<ComponentSet> {
+      return new Promise(resolve => resolve(new ComponentSet()));
+    }
+  }
+
+  beforeEach(async () => {
+    jest.spyOn(process, 'cwd').mockReturnValue(dummyProcessCwd);
+    jest.spyOn(fs, 'existsSync').mockReturnValue(true);
+    workspaceContextGetInstanceSpy.mockReturnValue(mockWorkspaceContext);
+    createSourceTrackingSpy.mockResolvedValue({} as any);
+    retrieveSpy.mockResolvedValue({ pollStatus: jest.fn() } as any);
+  });
+
+  it('should create Source Tracking before retrieving', async () => {
+    // Arrange
+    const executor = new TestRetrieveExecutor(
+      'testRetrieve',
+      'testRetrieveLog'
+    );
+    (executor as any).setupCancellation = jest.fn();
+
+    // Act
+    await (executor as any).doOperation(dummyComponentSet, {});
+
+    // Assert
+    expect(createSourceTrackingSpy).toHaveBeenCalled();
+    expect(retrieveSpy).toHaveBeenCalled();
+    const createSourceTrackingCallOrder =
+      createSourceTrackingSpy.mock.invocationCallOrder[0];
+    const retrieveCallOrder = retrieveSpy.mock.invocationCallOrder[0];
+    expect(createSourceTrackingCallOrder).toBeLessThan(retrieveCallOrder);
+  });
```

Comment:
for these two tests are there any error conditions worth checking?  None jump out at me but figured it was worth asking.

Reply:
I added one test here and one in the deployExecutor.test.ts file to cover the handling of SourceConflict errors.  Let me know if you think of any other scenarios that we should have a unit test for here.

Reply:
Trying to jump around in the code to see where else we're aware of potential errors, and it looks like the postconditionChecker is wrapping loadCache, createDiffs, and handleConflicts all in a try block. There don't look to be any tests with more helpful information though... Not sure if they were just being overly cautious, but might be something to consider.

https://github.com/forcedotcom/salesforcedx-vscode/blob/ea0dc1b133a2e25dfdbceec2aebd611c08154bca/packages/salesforcedx-vscode-core/src/commands/util/postconditionCheckers.ts#L237
https://github.com/forcedotcom/salesforcedx-vscode/blob/ea0dc1b133a2e25dfdbceec2aebd611c08154bca/packages/salesforcedx-vscode-core/test/vscode-integration/commands/util/postconditionCheckers.test.ts#L577